### PR TITLE
feat: add scores and evaluations as maps

### PR DESCRIPTION
**Pull request by JoanFM**
**Changes introduced**
It turns `scores` and `evaluations` inside `DocumentProto` into a `Map-like` struct to have the following experience:

```python
from jina import Document
d = Document()
d.scores['cosine'] = 1
d.scores['euclidean'] = 0.6
d.evaluations['precision'] = 0.5
d.evaluations['recall'] = 0.7
```

### CODE REVIEW

**Comment by bwanglzu** [comment]
one question, do we still need `NamedProto`?

**Comment by JoanFM** [comment]
> one question, do we still need `NamedProto`?

maybe we could remove, but it still convenient and not damaging I think

File: jina/types/score/map.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+from ..score import NamedScore
+from ..mixin import ProtoTypeMixin
+from ...excepts import BadNamedScoreType
+from ...helper import typename
+from ...proto import jina_pb2
+
+__all__ = ['MappedNamedScore']
+
+
+class MappedNamedScore(ProtoTypeMixin):
```
Comment (bwanglzu) [inline jina/types/score/map.py:14]: suggestion, refine the name to `NameScoreMapping`? Python community often use this kind of name  such as `Mapping`, `MutableMapping` etc..

File: jina/types/score/map.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+                raise ValueError(f'{typename(scores)} is not recognizable')
+        except Exception as ex:
+            raise BadNamedScoreType(
+                f'fail to construct a MappedNamedScore from {scores}'
+            ) from ex
+
+    def __setitem__(
+        self,
+        key: str,
+        value: Union[jina_pb2.NamedScoreProto, NamedScore, float, np.generic],
```
Comment (bwanglzu) [inline jina/types/score/map.py:69]: i think it's safe to remove `NamedScore` unless we really want to keep the fields other than `op_name` and `value`

Comment (JoanFM) [reply jina/types/score/map.py:69]: so far we want to keep them

**Review by bwanglzu — COMMENTED** [review]
left some comments

**Comment by bwanglzu** [comment]
and we might be able to use `collections.abc`, same as what we've used in `DocumentArray` (inherit from `MutableSequence`)

File: jina/types/score/map.py (hunk 2 above)
Comment (bwanglzu) [reply jina/types/score/map.py:69]: okay

File: jina/types/score/map.py (hunk 1 above)
Comment (JoanFM) [reply jina/types/score/map.py]: Good point!, Added

**Comment by JoanFM** [comment]
> and we might be able to use `collections.abc`, same as what we've used in `DocumentArray` (inherit from `MutableSequence`)， maybe [MutableMapping](https://docs.python.org/3/library/collections.abc.html#collections.abc.MutableMapping)

Added

**Review by bwanglzu — DISMISSED** [review]
Looks great to me! let's see of other people have some ideas!

**Comment by JoanFM** [comment]
Wait @bwanglzu I am gonna change the `cookbooks`

**Review by florian-hoenicke — APPROVED** [review]

**Review by hanxiao — APPROVED** [review]

**Pull request merged**